### PR TITLE
KAA-1167: Notifications - reviewed.

### PR DESCRIPTION
**Pull request by hunch7**

### CODE REVIEW

File: doc/Programming-guide/Key-platform-features/Notifications/index.md [hunk 1]
```diff
@@ -10,33 +10,44 @@ sort_idx: 70
...
-We also recommend that you review [collecting endpoint profiles guide]({{root_url}}Programming-guide/Key-platform-features/Data-collection) and [endpoint groups]({{root_url}}Programming-guide/Key-platform-features/Endpoint-groups) before you proceed with this guide.
+## Prerequisites
 
-### Basic architecture
+To use the examples below, you need to first set up either a [Kaa Sandbox]({{root_url}}Glossary/#kaa-sandbox) or a full-blown [Kaa cluster]({{root_url}}Glossary/#kaa-cluster).
+After that, you need to create a tenant and an application.
+To do this, you can use the [server REST API]({{root_url}}Programming-guide/Server-REST-APIs/#!/User/createKaaAdmin) or the [Administration UI]({{root_url}}Administration-guide/Users-management/#managing-tenant-admins).
```
Comment (Kirill380) [inline doc/Programming-guide/Key-platform-features/Notifications/index.md]: Wrong name of link [server REST API](should be tenant and user REST API) and  open a wrong page --  creating Kaa admin instead of creating tenant and  tenant admin [ /Programming-guide/Server-REST-APIs/#!/Tenant/editTenant and Programming-guide/Server-REST-APIs/#!/User/editUser]. And where is API for creating an application?

Comment (hunch7) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md]: Fixed.

File: doc/Programming-guide/Key-platform-features/Notifications/index.md [hunk 2]
```diff
@@ -52,66 +63,48 @@ The default notification schema installed for Kaa applications is empty. For the
...
-
-1. In the **Notification schemas** window for the application, click **Add schema**.
-2. In the **Add notification schema** window, create a notification schema either by using the [schema form]({{root_url}}Administration-guide/Tenants-and-applications-management/#schema-form) or by uploading a schema in the [Avro](http://avro.apache.org/docs/current/spec.html) format from a file.
-
-![Add Notification Schema 2](images/add_notification_schema_2.png)
-3. Click **Add** to save the schema.
-
-If you want to review the added Avro schema, open the corresponding **Notification schema** window by clicking the schema in the **Notification schemas** window.
-
-![Add Notification Schema 3](images/add_notification_schema_3.png)
```
Comment (Kirill380) [inline doc/Programming-guide/Key-platform-features/Notifications/index.md]: #L62

File: doc/Programming-guide/Key-platform-features/Notifications/index.md [hunk 3]
```diff
@@ -52,66 +63,48 @@ The default notification schema installed for Kaa applications is empty. For the
 }
 ```
 
-You can configure your own notification schema via [REST API]({{root_url}}Programming-guide/Server-REST-APIs/#!/Notifications/saveNotificationSchema) call or use Administration UI as shown below. 
-
-
-The list of notification schemas created by a tenant developer for the application is shown in the **Notification schemas** window.
+To configure your own notification schema, use the [REST API]({{root_url}}Programming-guide/Server-REST-APIs/#!/Notifications/saveNotificationSchema) or select it from the **Notification** page under the **Schemas** section of the application.
```
Comment (Kirill380) [inline doc/Programming-guide/Key-platform-features/Notifications/index.md]: [ss](#L62)

File: doc/Programming-guide/Key-platform-features/Notifications/index.md [hunk 4]
```diff
@@ -52,66 +63,48 @@ The default notification schema installed for Kaa applications is empty. For the
 }
 ```
 
-You can configure your own notification schema via [REST API]({{root_url}}Programming-guide/Server-REST-APIs/#!/Notifications/saveNotificationSchema) call or use Administration UI as shown below. 
-
-
-The list of notification schemas created by a tenant developer for the application is shown in the **Notification schemas** window.
```
Comment (Kirill380) [inline doc/Programming-guide/Key-platform-features/Notifications/index.md:58]: Why did you remove this paragraph?  One describe the image below.

File: doc/Programming-guide/Key-platform-features/Notifications/index.md (hunk 3 above)
Comment (Kirill380) [inline doc/Programming-guide/Key-platform-features/Notifications/index.md]: Move after add_notification_schema_1.png image and combine with the paragraph on line number 70. Remove **or select it from the **Notification** page under the **Schemas** section of the application.**

File: doc/Programming-guide/Key-platform-features/Notifications/index.md (hunk 4 above)
Comment (hunch7) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md:58]: Because it's too obvious even for unexperienced users. When we create a new item, of course we expect it to appear in the list of the same type items. It doesn't deserve a paragraph and a screenshot. Plus, it's redundant because the 'Notification schema' section contains the screenshot of that list.

Comment (hunch7) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md:58]: This screenshot is given to illustrate this portion of the sentence above it: "select it from the Notification page under the Schemas section of the application".

File: doc/Programming-guide/Key-platform-features/Notifications/index.md (hunk 3 above)
Comment (hunch7) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md]: My logic is this: I first say that users can configure the existing schemas and give the screenshot to show where they can do it. Then I say that users can create their own schemas, but instead of writing the procedure, I give the link to the similar procedure of creating a clien-side profile schema.

File: doc/Programming-guide/Key-platform-features/Notifications/index.md [hunk 5]
```diff
@@ -151,45 +144,38 @@ Assuming that you have created custom endpoint groups from the [Using endpoint g
...
-Notification pipelines manage individual notifications within a topic. A notification remains queued in the pipeline until its time-to-live (TTL) expires, after which the notification is dropped. The notification pipeline type specifies the scope of notifications delivery. 
-It can be either multicast (targeted to an unbounded number of endpoints) or unicast (targeted to a single specific endpoint).
-    
-**Multicast pipelines** 
-<br/>
-Whenever a notification is sent to a topic, it gets added to the corresponding pipeline with a unique sequential index per pipeline. Endpoints independently maintain their position in every pipeline by remembering the last sequential index per pipeline that they received.
-    
-**Unicast pipelines** 
-<br/>
-Whenever a notification is sent to a topic, it gets added to the corresponding pipeline with a unique ID. Endpoints independently maintain their pipeline by reporting all received notification IDs. The server removes the notification from the pipeline once it receives a receipt confirmation from the endpoint.
```
Comment (Kirill380) [inline doc/Programming-guide/Key-platform-features/Notifications/index.md:184]: Same question as above one

File: doc/Programming-guide/Key-platform-features/Notifications/index.md [hunk 6]
```diff
@@ -151,45 +144,38 @@ Assuming that you have created custom endpoint groups from the [Using endpoint g
...
+A notification will be queued for delivery until the time you specified in the **Expires at** field of the **Notification details** page.
+If you leave this field blank, the message will be queued until it is delivered.
+
+If you specified an endpoint ID in the **KeyHash** field, the notification will be only sent to that specific endpoint.
+If you leave this field blank, the notification will be sent to all endpoints subscribed to the selected notification topic.
 
-Notification pipelines manage individual notifications within a topic. A notification remains queued in the pipeline until its time-to-live (TTL) expires, after which the notification is dropped. The notification pipeline type specifies the scope of notifications delivery. 
-It can be either multicast (targeted to an unbounded number of endpoints) or unicast (targeted to a single specific endpoint).
-    
-**Multicast pipelines**
```
Comment (Kirill380) [inline doc/Programming-guide/Key-platform-features/Notifications/index.md:178]: Why did you delete this section? It describes in details how pipelines multicast work.

Comment (hunch7) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md:178]: I replaced it with more user-oriented content. In reality, user sees two fields -- 'Expires at' and 'KeyHash'. I explained where they are and what they do in plain language. There is no need to explain all the low-level logic behind those fields and introduce complex theoretical terminology that the user will never stumble upon using our product. Users don't need (and most probably don't want) this kind of information, it confuses them.

File: doc/Programming-guide/Key-platform-features/Notifications/index.md (hunk 5 above)
Comment (hunch7) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md:184]: See my comment above.

File: doc/Programming-guide/Key-platform-features/Notifications/index.md (hunk 3 above)
Comment (Kirill380) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md]: User can't configure the existing schemas, one just creates it via REST API or Admin UI. You separate REST API creation from creation via Admin UI. 'Configure' in this context means create.

File: doc/Programming-guide/Key-platform-features/Notifications/index.md (hunk 4 above)
Comment (Kirill380) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md:58]: Ok, then better move this image to the line 72 as the result of creation

File: doc/Programming-guide/Key-platform-features/Notifications/index.md (hunk 1 above)
Comment (Kirill380) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md]: I don't see changes, Did you push your fix commit?

Comment (hunch7) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md]: Yes, I rephrased the sentence and included the right links.

File: doc/Programming-guide/Key-platform-features/Notifications/index.md (hunk 4 above)
Comment (hunch7) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md:58]: Done.

File: doc/Programming-guide/Key-platform-features/Notifications/index.md (hunk 3 above)
Comment (hunch7) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md]: You're right, they can configure data values, not schemas. The word 'configure' should never be used to mean 'create'. Sentence rephrased.

File: doc/Programming-guide/Key-platform-features/Notifications/index.md [hunk 7]
```diff
@@ -151,45 +142,38 @@ Assuming that you have created custom endpoint groups from the [Using endpoint g
...
-The file with the following contents will match the default sandbox notification schema.
+Below is an example of the uploaded file contents that will match the default Sandbox notification schema.
 
 ```
 {"message": "Hello from Kaa!"}
 ```
 
-#### Notification pipelines
+A notification will be queued for delivery until the time you specified in the **Expires at** field of the **Notification details** page.
+If you leave this field blank, the message will be queued until it is delivered.
```
Comment (ashvayka) [inline doc/Programming-guide/Key-platform-features/Notifications/index.md]: This is not true, default notification ttl is 1 week. This is hardcoded parameter in NotificationServiceImpl. And yes, i will open a ticket for this :-)

Comment (ashvayka) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md]: http://jira.kaaproject.org/browse/KAA-1431

File: doc/Programming-guide/Key-platform-features/Notifications/index.md (hunk 6 above)
Comment (ashvayka) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md:178]: @hunch7 although your description is quite simple and clear to regular software product user, out product and this documentation is written for software engineers(SE). As a SE, i would like to know about this pipelines and is interested in this details. However, your point of view is also very logical. There are different types of SEs. Some of them are concentrated on client-side code and may be not interested in such details. 
I suggest to keep your version and then add smth like "See Notification Pipelines section for more details". Notification pipelines section will contain old version.

File: doc/Programming-guide/Key-platform-features/Notifications/index.md [hunk 8]
```diff
@@ -618,10 +605,11 @@ error_code = kaa_sync_topic_subscriptions(kaa_client_get_context(kaa_client)->no
...
 </div></div>
 
+### Default notification listener
 
-#### Default notification listener
-There are two types of topic notification listeners: the default and topic specific. To receive notifications, add at least one default listener (the number of default listeners is not limited) as shown in the following code block. 
-As a result, the listener will receive notifications from all topics (mandatory topics, as well as optional topics having been subscribed to). 
+There are two types of topic notification listeners: default and topic-specific.
+To receive notifications, add at least one default listener.
+As a result, the listener will receive notifications from all topics (all mandatory topics and all optional topics that the endpoint group is subscribed to.
```
Comment (ashvayka) [inline doc/Programming-guide/Key-platform-features/Notifications/index.md]: missing ")" symbol?

**Comment by Kirill380** [comment]
LGTM from my side

File: doc/Programming-guide/Key-platform-features/Notifications/index.md (hunk 8 above)
Comment (hunch7) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md]: Fixed.

File: doc/Programming-guide/Key-platform-features/Notifications/index.md (hunk 6 above)
Comment (hunch7) [reply doc/Programming-guide/Key-platform-features/Notifications/index.md:178]: Fixed.
